Task 1: Develop a new Web Application

Brunoara12 opened the pull request:
1.	Home/Landing page
2.	Second page with navigation
3.	Traditional HTML Forms (2) 
  3.1.	Required field
  3.2.	Optional field
4.	Synchronous call (Java MVC)
5.	Asynchronous call (AngularJS Service to Rest-style Java Controller)

Brunoara12 committed:
First Commit

Brunoara12 committed:
Merge branch 'dev' of https://github.com/Brunoara12/schedule-now into dev

Brunoara12 committed:
Removed Ignored Files

Brunoara12 committed:
task1 feat: - added update and delete CRUD operations for asynchronous calls to REST Controller
- updated scheduling view page to match the calls

Brunoara12 committed:
hotfix: - cleaned up code
- removed test pages

minh-nguyet-nguyen commented on src/main/webapp/WEB-INF/web.xml:
```diff
@@ -0,0 +1,18 @@
+<web-app>
+ 	<display-name>Archetype Created Web Application</display-name>
+  
+<!-- 	<servlet> -->
```
You should probably get rid of these commented lines if you haven't already

minh-nguyet-nguyen commented on src/main/java/com/bruno/controller/SchedulingController.java:
```diff
@@ -0,0 +1,70 @@
+		Job newJob = jobService.update(id, job);
+		
+		return newJob != null ? new ResponseEntity<Job>(job, HttpStatus.OK) : 
+			new ResponseEntity<Job>(HttpStatus.NOT_MODIFIED);
```
This is probably fine for now, but we've been told to usually keep our HTTP statuses to anything in the 200s and 500s

chriswbluer commented on src/main/java/com/bruno/controller/SchedulingController.java:
```diff
@@ -0,0 +1,70 @@
+	
+		List<Job> jobs = jobService.getAllJobs();
+		
+		return jobs.isEmpty() ? new ResponseEntity<List<Job>>(HttpStatus.NO_CONTENT) :
```
This is awesome. I didn't know Java had ternary operators.

Brunoara12 committed:
hotfix:
- removed duplicated pom dependencies
- removed deprecated inheritance
- added synchronous form submit login from index page
- updated PUT request response status code from 300 -> 200
- jsp and css format updates

Mbd06b commented on pom.xml:
```diff
@@ -0,0 +1,78 @@
+  	<!-- https://mvnrepository.com/artifact/org.springframework/spring-webmvc -->
+	<dependency>
+	    <groupId>org.springframework</groupId>
+	    <artifactId>spring-webmvc</artifactId>
+	    <version>5.1.8.RELEASE</version>
```
Nice, no more duplicates.

Mbd06b commented on src/main/java/com/bruno/controller/HomeController.java:
```diff
@@ -0,0 +1,43 @@
+		ModelAndView mv = new ModelAndView(); 
+		mv.setViewName("index");
+		
+		return mv;	}
```
close bracket to next line.  Add Sonarlint plugin to your IDE

Mbd06b commented on src/main/java/com/bruno/controller/HomeController.java:
```diff
@@ -0,0 +1,43 @@
+
+	@PostMapping("/login")
+	public ModelAndView submit(@RequestParam("username") String username, 
+			@RequestParam("password") String password) {
```
Don't put parameterize values which could contain PII, default to using a @RequestBody

Mbd06b commented on src/main/java/com/bruno/models/Job.java:
```diff
@@ -0,0 +1,92 @@
+	private String date;
+	private String description;
+	
+	public Job() {
+		this.id = 0L;
+	}
+	
+	public Job(Long id, String customerName, String address, String email, String phone, 
+			String date, String description) {
+		this.id = id;
+		this.customerName = customerName;
+		this.address = address;
+		this.email = email;
+		this.phone = phone;
+		this.date = date;
+		this.description = description;
```
Unless you're really using these, I suggest not defining a null/custom constructors and let Spring handle the default until you need to implement an interface-based copy constructor.

Mbd06b commented on src/main/java/com/bruno/models/Job.java:
```diff
@@ -0,0 +1,92 @@
+
+	@Override
+	public String toString() {
+		return "Job [id=" + id + ", customerName=" + customerName + ", address=" + address + ", email=" + email
+				+ ", phone=" + phone + ", date=" + date + ", description=" + description + "]";
+	}
```
don't Override toString to display PII properties.

Mbd06b commented on src/main/webapp/WEB-INF/web.xml:
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE web-app PUBLIC
+ "-//Sun Microsystems, Inc.//DTD Web Application 2.3//EN"
+ "http://java.sun.com/dtd/web-app_2_3.dtd" >
```
web-app_2_3?  what is DTD and how does that compare to other Java web-app deployment descriptors?  Is this an old webapp version? be careful with this value, because it can change how you have to load your application context when trying to spin up compontents for later unit testing.

Mbd06b commented on src/main/webapp/static/js/controllers/scheduling_controller.js:
```diff
@@ -0,0 +1,100 @@
+	function getAllJobs() {
+		SchedulingService.getAllJobs()
+		.then(
+			function(e) {
```
spell out the variable here, it looks like your normal return is an execption. This 'e'  should be something like "data"

Mbd06b commented on src/main/webapp/static/js/controllers/scheduling_controller.js:
```diff
@@ -0,0 +1,100 @@
+	
+	function update(id){
+		$log.log("job: " + id + " will be edited")
+		for(let i = 0; i < vm.jobs.length; i++){
+			if(vm.jobs[i].id === id) {
+				vm.job = angular.copy(vm.jobs[i])
```
You don't need this for-loop to select an existing job in a list, just past the model present the the row in from ng-repeat. and perform the anuglar.copy() on it to set your form subject.

Mbd06b commented on src/main/webapp/static/js/controllers/scheduling_controller.js:
```diff
@@ -0,0 +1,100 @@
+	}
+	
+	function reset() {
+		vm.job = {id: null, customerName: '', address: '', email: '',
+		phone: '', data: '', description: ''}
```
just vm.job = {} should suffice.  These explicit references are not a scalable pattern, and are not useful documentation.

Mbd06b commented on src/main/webapp/static/js/services/scheduling_service.js:
```diff
@@ -0,0 +1,69 @@
+SchedulingService.$inject = ['$http', '$log']
+
+function SchedulingService($http, $log) {
+	var REST_SERVICE_URI = 'http://localhost:8080/ScheduleNow/scheduling/jobs/';
```
don't ever hardcode the full protocol//host:port/context,    Only ever use the relative path  "scheduling/jobs"

Mbd06b commented on src/main/webapp/static/js/services/scheduling_service.js:
```diff
@@ -0,0 +1,69 @@
+				return res.data;
+			},
+			function (err) {
+				$log.error('Error while fetching Jobs ', err);
```
good use of $log methods provided by anugarjs

Mbd06b commented on src/main/webapp/static/js/controllers/scheduling_controller.js:
```diff
@@ -0,0 +1,100 @@
+ 'use strict';
+ 
+angular
+	.module('myApp')
+	.controller('SchedulingController', SchedulingController) 
+
+SchedulingController.$inject = ['SchedulingService','$log']
+
+function SchedulingController(SchedulingService, $log){
```
very nice John Papa style declarations

Mbd06b commented on src/main/webapp/views/scheduling.jsp:
```diff
@@ -0,0 +1,152 @@
+<%@ page language="java" contentType="text/html; charset=ISO-8859-1" pageEncoding="ISO-8859-1" isELIgnored="false"%>
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core"  %>
+<html>
```
508 error missing default lang attribute. 
I'm seeing a lot of missing meta here.  Run your whole loaded browser source through https://validator.w3.org 
and use Axe Devtools to perform a full 508 accessibility compliance check on your webpage.  
https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US 
Fix all errors.

Mbd06b commented on src/main/webapp/views/scheduling.jsp:
```diff
@@ -0,0 +1,152 @@
+<%@ page language="java" contentType="text/html; charset=ISO-8859-1" pageEncoding="ISO-8859-1" isELIgnored="false"%>
```
conform pageEncoding to UTF-8

Mbd06b commented on src/main/webapp/views/scheduling.jsp:
```diff
@@ -0,0 +1,152 @@
+          </div>
+      </div>
+       
+     	<script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.6.9/angular.js"></script>
```
move AngularJS CDN library '<script>' tag to the '<head>' element.

Mbd06b reviewed (changes requested):
Run your running webapp through w3 html validator, and Axe Devtools for accessibility. adjust your Controller to pass the JSON model, rather than performing secondary iteration.

Brunoara12 committed:
hotfix: - updated to validate html and make it 508 compliant
- several cleanup code fixes
